Better support for emoji wcwidth

u-quark opened the pull request:
Change the implementation of `wcwidth` in vty taking the implementation from [musl](http://git.musl-libc.org).

 * Make `wcwidth` compatible with Unicode 12.1.0.
 * Add code that generates the C tables from the Unicode data, also from [musl](https://github.com/richfelker/musl-chartable-tools).
 * Add a utility executable that runs through the Unicode codepoints and compares what `wcwidth` thinks with the actual implementation of the terminal.
 * Add a demo that prints a bunch of emoji character combinations from various versions of Unicode along with the width calculated by the new implementation of `wcwidth`.
 * Fix some test cases that deal with wide characters.

I think, the new implementation is simpler and more efficient although, I have to admit, that I don't understand all the bitwise logic.

I've tested this on a few terminals on Linux with overall good results.

This should bring us one step closer in closing #175.

u-quark committed:
Add demo exec to show emoji and their widths

u-quark committed:
New implementation for wcwidth

This implementation is using a two-level lookup table.

This new implementation follows Unicode version 12.1.0.

u-quark committed:
Add unicode tools

This code generates tables from Unicode data that can then be used in C
code to efficiently access the Unicode data.

u-quark committed:
Fix tests for wide characters

The previous quickcheck code, when picking double column characters
(wide) it was including ambiguous or narrow characters like:

 * 3248 CIRCLED NUMBER TEN ON BLACK SQUARE
 * 4DC0 HEXAGRAM FOR THE CREATIVE HEAVEN

See https://www.unicode.org/Public/12.0.0/ucd/EastAsianWidth.txt for
more details.

Also, included more characters in the class of wide characters.

u-quark committed:
Add wcwidth tester executable

This executable goes through the Unicode code-points and figures out how
many columns each character takes in the terminal. The result is written
at "output.txt".

u-quark committed:
Add attribution notice for musl code

u-quark commented:
I could not make `cabal sdist` include the C header files under `unicode_tools` by mentioning the files in `install-includes` in the executable. For whatever reason I had to mention the files in the library as well: https://github.com/jtdaugherty/vty/pull/179/commits/79955363e056fd42ecf8517bfc03780f8a95390e
Not sure if there is a better way of doing this.

jtdaugherty commented:
@u-quark thanks so much for working on this patch. I'm sorry I haven't had time to look at it in depth until now. I took a look and the patch definitely improves emoji handling in Matterhorn, a project I work on. I'm curious if @glguy can see how this impacts his IRC client.

jtdaugherty commented:
A thing I found, after seeing this tool run for a few seconds:
```
$ vty-wcwidth-tester
00d800: vty-wcwidth-tester: <stdout>: hPutChar: invalid argument (invalid character)
```

jtdaugherty commented:
Also: I tried running `vty-emoji-demo`. When I ran it, nothing appeared in the terminal. I checked `top` and the program was using 100% of a CPU and steadily consuming more RAM. I killed it when it got to about 8 GB of residency.

u-quark commented:
@jtdaugherty Thanks for looking at this!

> $ vty-wcwidth-tester
00d800: vty-wcwidth-tester: <stdout>: hPutChar: invalid argument (invalid character)

I can confirm that I get the same error. I should have filtered out more character categories:
```diff
 shouldConsider c =
   case generalCategory c of
     Control     -> False
+    NotAssigned -> False
+    Surrogate   -> False
     _           -> True
```
I will fix!

`vty-emoji-demo` consuming 100% CPU and consuming memory is more surprising as the program is very simple and is supposed to print a fixed length text and exit upon pressing `Esc`, as the other demo in the code of vty.

Could you tell me the versions of a few components to try and narrow the bug down and reproduce it: versions of distribution, terminfo library, terminal emulator and ghc compiler. Could you also tell me the encoding/locale you are using? Do you have any idea which part of the code consumes 100% CPU and what it allocates in memory?

jtdaugherty commented:
I did some digging and it looks like the problem arises in `EmojiDemo.hs` when `showEmoji` goes to call `wcswidth` on a particular value of `e`. The value is `"\128992"` and the value of `wcswidth e` is `4294967295`, which results in `replicate (wcswidth e) '.'` and the corresponding CPU/memory behavior.

That value is `"\x1F7E0" -- Orange Circle`. Even if I comment that one out, others also cause this issue for me.

* Distribution: macOS Catalina 10.15.2
* Terminfo: `terminfo-0.4.1.2`
* Terminal emulator: iTerm2 3.3.7
* GHC: 8.6.5

u-quark commented:
Ah I see. Not so fixed length string after all! I will have a look and try to fix. It looks like some integer overflow, interpreting signed int -1 as an unsigned int. Thanks for looking into this.

u-quark commented:
@jtdaugherty I fixed the two bugs you mentioned and squashed/refactored some of the commits.

I am a bit confused though as to why `wcswidth "\128992"` would return `4294967295`/`-1` in your case. It shouldn't. Can you check that with the new code you get the correct value (`2`)?

jtdaugherty commented:
The fixes seem to have improved things. Both the width tester and emoji demo programs now seem to work. Thanks!

I do want to mention that I get some junk output from some of the wcwidth tests. It looks like this:
```
$ vty-wcwidth-tester
02fa1d: 𪘀^[[36;11R
```
(The output at the end of the line after the wide character changes periodically as the program runs through the tests.)

u-quark commented:
This is probably ok as long as you get a sensible `output.txt` file at the end. It probably has to do with how the terminal interprets the escape sequences. Can you share with me this file that got generated on your machine after the program finished? `vty-wcwidth-tester` is supposed to do something very similar to the code from @glguy: https://gist.github.com/glguy/a802eaaac5a00ab41e3c3ca33507634b

jtdaugherty commented:
Closing as per my comment on https://github.com/jtdaugherty/vty/issues/175. Thank you again for your work on this!

jtdaugherty commented on unicode_tools/WcwidthTester.hs:
```diff
@@ -0,0 +1,68 @@
+         if isEmoji i
+           then do
+             checkString onlyInconsistencies h [i, '\xFE0E']
+             checkString onlyInconsistencies h [i, '\xFE0F']
```
@u-quark Could you shed some light on what this `isEmoji` test is for? I don't know what these extra `checkString` calls are for. Why isn't this loop body just one call to `checkString`?

u-quark replied:
So this is fun! https://www.unicode.org/reports/tr51/tr51-16.html#Presentation_Style

Unicode defines two characters called presentation selectors `'\xFE0E'` and `'\xFE0F'`. They influence if the character is going to be presented as an emoji (with color) or text. For at least one terminal emulator that I have tested (specifically Kitty) if these characters exist after an emoji they affect how many columns the emoji is taking up.

glguy commented on src/Graphics/Text/Width.hs:
```diff
@@ -14,10 +14,14 @@ module Graphics.Text.Width
 import qualified Data.Text.Lazy as TL
 
-foreign import ccall unsafe "vty_mk_wcwidth" wcwidth :: Char -> Int
+foreign import ccall unsafe "wcwidth" wcwidth_ :: CWchar -> CInt
```
It's safer to prefix the publicly visible names with something less likely to clash with any other library that this might be used with. These names have to be unique across all libraries linked into an application which makes the `vty_` prefix attractive.

u-quark replied:
This is a good idea!

u-quark replied:
Especially because it seems like `wcwidth` already exists in general. I pushed a fix on https://github.com/u-quark/vty/commits/fix-for-better-support-for-emoji-width but this PR does not update because it is closed.
